pkg/blobcache/blobcache.go

```diff
@@ -0,0 +1,410 @@
+	return d.destination.DesiredLayerCompression()
+}
+
+func (d *blobCacheDestination) AcceptsForeignLayerURLs() bool {
+	return d.destination.AcceptsForeignLayerURLs()
+}
+
+func (d *blobCacheDestination) MustMatchRuntimeOS() bool {
+	return d.destination.MustMatchRuntimeOS()
+}
+
+func (d *blobCacheDestination) IgnoresEmbeddedDockerReference() bool {
+	return d.destination.IgnoresEmbeddedDockerReference()
+}
+
+func (d *blobCacheDestination) PutBlob(ctx context.Context, stream io.Reader, inputInfo types.BlobInfo, isConfig bool) (types.BlobInfo, error) {
+	if inputInfo.Digest != "" && len(d.reference.directories) > 0 {
+		if d.reference.removeFirstDirectoryOnCleanup && d.reference.directories[0] == "" {
+			temp, err := ioutil.TempDir("", "")
+			if err != nil {
+				return types.BlobInfo{}, errors.Wrapf(err, "error creating new temporary directory for blob cache")
+			}
+			d.reference.directories[0] = temp
+		}
+		filename := filepath.Join(d.reference.directories[0], makeFilename(inputInfo.Digest, isConfig))
+		f, err := os.OpenFile(filename, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0600)
+		if err == nil {
+			stream = io.TeeReader(stream, f)
+			defer f.Close()
+		}
+	}
+	return d.destination.PutBlob(ctx, stream, inputInfo, isConfig)
+}
+
+func (d *blobCacheDestination) HasBlob(ctx context.Context, info types.BlobInfo) (bool, int64, error) {
+	present, size, err := d.reference.HasBlob(info)
+	if err != nil {
+		return false, -1, errors.Wrapf(err, "error checking for blob %q in cache %#v", info.Digest.String(), d.reference.directories)
+	}
+	if present {
+		return present, size, nil
+	}
+	return d.destination.HasBlob(ctx, info)
+}
+
+func (d *blobCacheDestination) ReapplyBlob(ctx context.Context, info types.BlobInfo) (types.BlobInfo, error) {
+	present, _, err := d.destination.HasBlob(ctx, info)
+	if err != nil {
+		return types.BlobInfo{}, errors.Wrapf(err, "error checking for blob %q in cache %#v", info.Digest.String(), d.reference.directories)
+	}
+	if present {
+		return d.destination.ReapplyBlob(ctx, info)
+	}
+	for _, directory := range d.reference.directories {
+		if directory == "" {
+			continue
+		}
+		for _, isConfig := range []bool{false, true} {
+			filename := filepath.Join(directory, makeFilename(info.Digest, isConfig))
+			f, err := os.Open(filename)
+			if err == nil {
+				defer f.Close()
+				return d.destination.PutBlob(ctx, f, info, isConfig)
+			}
+		}
+	}
+	return d.destination.ReapplyBlob(ctx, info)
+}
+
+func (d *blobCacheDestination) PutManifest(ctx context.Context, manifestBytes []byte) error {
+	if len(d.reference.directories) > 0 {
+		manifestDigest, err := manifest.Digest(manifestBytes)
+		if err != nil {
+			logrus.Warnf("error digesting manifest %q: %v", string(manifestBytes), err)
+		} else {
+			filename := filepath.Join(d.reference.directories[0], makeFilename(manifestDigest, false))
+			if err = ioutil.WriteFile(filename, manifestBytes, 0600); err != nil {
```

Comment:
Does anything read these cached manifests?

Reply:
I could see us wanting to cache manifests for non-`nil` instanceDigest values.  Added a check for a cached copy when an `instanceDigest` is supplied to `GetManifest()`.